### ISSUE TITLE
Replace Symfony ClassLoader by composer ClassLoader

### ISSUE THREAD

**Issue by mcg-web** [overblog/GraphQLBundle#189]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Version/Branch   | 0.9.x

Symfony ClassLoader is deprecated and should be replace by composer autoloader. In Symfony 4 this bundle generated schema types will be persist, so maybe just using the `autoload` composer.json entry will be sufficient and more optimized.

**Comment by stof** [comment]
If you need your own autoloader, you could indeed instantiate a composer ClassLoader (people using a Symfony bundle without having composer are probably too edge case to require supporting this).
Just make sure you don't alter the configuration of the instance available as `$loader = require 'vendor/autoload.php'`. This would break several optimizations done in Composer for the generated class loader. Use a second instance instead.

Relying on the generated Composer autoloader configured through `composer.json` depends on when the classes are generated (if it is too late, it breaks the usage of classmap-authoritative autoloaders). And it may require the user to alter their composer.json.

**Comment by stof** [comment]
If types are generated as classes meant to be committed in the repo, using the composer autoloader is totally fine.
But make sure that the generated files are located in a way allowing to use PSR-4 autoloading (otherwise, dev environments will be a pain if they have to redump the composer classmap everytime they add a new type)

**Comment by mcg-web** [comment]
First, thanks @stof for your help.

Yes the generated types PSR-4 ready combine with a classMap File that is keep up to date.  
Using my own Composer ClassLoader look to be the best way to go, as altering composer.json could be more painful for users... 

Can this example do it?

```php
$loader = new \Composer\Autoload\ClassLoader();
$classMap = 'my/path-to/__classes.map'
$loader->addClassMap($classMap);
// a little optimization
$loader->setClassMapAuthoritative(false);

$loader->register();
```

Thank you again

**Comment by stof** [comment]
@mcg-web yes (except that `$loader->setClassMapAuthoritative(false);` is useless as `false` is the default value).

However, if the generated types are meant to be committed in the project (rather than being generated on cache warmup for instance), you may not need this at all (just instructing users to add the autoloading rule for their types in composer.json)

**Comment by mcg-web** [comment]
Ok, need some more changes and documentation to make types committed, we will take the change of @renatomefi in first place. Thank you again @stof and also @renatomefi...

**Comment by renatomefi** [comment]
Maybe we could do this in 2 steps, first one just replacing the load, and secondly replacing the whole strategy of the custom types!

**Comment by mcg-web** [comment]
I start working on step 2 @renatomefi by giving possibility to disabled the ClassLoader Listener so doing allowing custom autoloader.

**Comment by renatomefi** [comment]
@mcg-web sounds great! :+1:

**Comment by mcg-web** [comment]
both steps are implemented and will be release in 0.9, thank you for helping me on this :+1:
